File: src/main/services/patterns.ts
/**
 * Infer and preserve field formats from user samples + history.
 * Dates stay dates; numbers keep shape (decimals, padding, currency); strings learn structure.
 */

export type InferredKind =
  | 'bool'
  | 'int'
  | 'int-padded'
  | 'float'
  | 'currency'
  | 'percent'
  | 'date-iso'
  | 'datetime-iso'
  | 'date-us'
  | 'date-eu'
  | 'date-slash-ymd'
  | 'email'
  | 'uuid'
  | 'phone'
  | 'alpha'
  | 'alnum'
  | 'string'

export interface FieldPattern {
  kind: InferredKind
  samples: string[]
  /** Integer padding width (e.g. 00123 → 5) */
  padWidth?: number
  minNum?: number
  maxNum?: number
  /** Decimal places for floats / currency */
  decimals?: number
  minLen: number
  maxLen: number
  /** Currency prefix like "$" */
  currencyPrefix?: string
  /** Whether currency uses thousands separators */
  useThousands?: boolean
}

const ISO_DATE = /^\d{4}-\d{2}-\d{2}$/
const ISO_DATETIME =
  /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{1,3})?(Z|[+-]\d{2}:\d{2})?$/
const DATE_US = /^(0?[1-9]|1[0-2])\/(0?[1-9]|[12]\d|3[01])\/\d{4}$/
const DATE_EU = /^(0?[1-9]|[12]\d|3[01])\/(0?[1-9]|1[0-2])\/\d{4}$/
const DATE_YMD_SLASH = /^\d{4}\/(0?[1-9]|1[0-2])\/(0?[1-9]|[12]\d|3[01])$/
const EMAIL = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
const UUID =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i
const PHONE = /^[+]?[\d\s().-]{7,20}$/
const INT = /^-?\d+$/
const FLOAT = /^-?\d+\.\d+$/
const CURRENCY = /^(\$|€|£)?-?\d{1,3}(,\d{3})*(\.\d+)?$|^(\$|€|£)?-?\d+(\.\d+)?$/
const PERCENT = /^-?\d+(\.\d+)?%$/
const ALPHA = /^[A-Za-z]+$/
const ALNUM = /^[A-Za-z0-9_-]+$/

function uniqueNonEmpty(values: string[]): string[] {
  return Array.from(new Set(values.map((v) => v.trim()).filter(Boolean)))
}

function allMatch(samples: string[], re: RegExp): boolean {
  return samples.length > 0 && samples.every((s) => re.test(s))
}

function decimalPlaces(s: string): number {
  const m = s.replace(/%$/, '').match(/\.(\d+)/)
  return m ? m[1].length : 0
}

function parseLooseNumber(s: string): number {
  return Number(s.replace(/[$€£,%]/g, '').replace(/,/g, ''))
}

/**
 * Detect pattern from history + the field's sampleValue (sample wins as format template).
 */
export function detectPattern(history: string[], sampleValue?: string): FieldPattern {
  const sample = sampleValue?.trim()
  const pool = uniqueNonEmpty([...(sample ? [sample] : []), ...history])

  if (pool.length === 0) {
    return { kind: 'string', samples: [], minLen: 4, maxLen: 12 }
  }

  // Prefer detecting from sample alone when present (user-defined format source of truth)
  const detectFrom = sample ? [sample] : pool
  const lens = pool.map((s) => s.length)
  const minLen = Math.min(...lens)
  const maxLen = Math.max(...lens)

  if (allMatch(detectFrom, /^(true|false)$/i) || allMatch(pool, /^(true|false)$/i)) {
    return { kind: 'bool', samples: pool, minLen: 4, maxLen: 5 }
  }

  if (allMatch(detectFrom, ISO_DATE) || (allMatch(pool, ISO_DATE) && !sample)) {
    return { kind: 'date-iso', samples: pool, minLen: 10, maxLen: 10 }
  }

  if (allMatch(detectFrom, ISO_DATETIME) || allMatch(pool, ISO_DATETIME)) {
    return { kind: 'datetime-iso', samples: pool, minLen: 19, maxLen: 30 }
  }

  // US vs EU: if sample matches one, lock that
  if (sample && DATE_US.test(sample) && !DATE_EU.test(sample)) {
    return { kind: 'date-us', samples: pool, minLen: 8, maxLen: 10 }
  }
  if (sample && DATE_EU.test(sample) && !/^\d{4}/.test(sample)) {
    // Prefer EU when day > 12 in samples, else if sample is DD/MM style from user region
    const day = Number(sample.split('/')[0])
    if (day > 12 || DATE_EU.test(sample)) {
      return { kind: 'date-eu', samples: pool, minLen: 8, maxLen: 10 }
    }
  }
  if (allMatch(detectFrom, DATE_YMD_SLASH) || allMatch(pool, DATE_YMD_SLASH)) {
    return { kind: 'date-slash-ymd', samples: pool, minLen: 8, maxLen: 10 }
  }
  if (allMatch(detectFrom, DATE_US) || allMatch(pool, DATE_US)) {
    return { kind: 'date-us', samples: pool, minLen: 8, maxLen: 10 }
  }
  if (allMatch(detectFrom, DATE_EU) || allMatch(pool, DATE_EU)) {
    return { kind: 'date-eu', samples: pool, minLen: 8, maxLen: 10 }
  }

  if (allMatch(detectFrom, EMAIL) || allMatch(pool, EMAIL)) {
    return { kind: 'email', samples: pool, minLen, maxLen }
  }

  if (allMatch(detectFrom, UUID) || allMatch(pool, UUID)) {
    return { kind: 'uuid', samples: pool, minLen: 36, maxLen: 36 }
  }

  if (allMatch(detectFrom, PERCENT) || allMatch(pool, PERCENT)) {
    const nums = pool.map(parseLooseNumber)
    const decimals = Math.max(...pool.map(decimalPlaces), 0)
    return {
      kind: 'percent',
      samples: pool,
      minLen,
      maxLen,
      minNum: Math.min(...nums),
      maxNum: Math.max(...nums),
      decimals
    }
  }

  if (allMatch(detectFrom, CURRENCY) || allMatch(pool, CURRENCY)) {
    const prefixMatch = (sample || pool[0]).match(/^(\$|€|£)/)
    const useThousands = pool.some((s) => s.includes(','))
    const nums = pool.map(parseLooseNumber)
    const decimals = Math.max(...pool.map(decimalPlaces), useThousands ? 2 : 0)
    return {
      kind: 'currency',
      samples: pool,
      minLen,
      maxLen,
      minNum: Math.min(...nums),
      maxNum: Math.max(...nums),
      decimals: decimals || 2,
      currencyPrefix: prefixMatch?.[1] ?? '$',
      useThousands
    }
  }

  // Padded integers (leading zeros) — keep as formatted string output
  if (sample && INT.test(sample) && /^-?0\d+/.test(sample)) {
    const nums = pool.filter((s) => INT.test(s)).map(Number)
    return {
      kind: 'int-padded',
      samples: pool,
      minLen,
      maxLen,
      padWidth: sample.replace(/^-/, '').length,
      minNum: nums.length ? Math.min(...nums) : 0,
      maxNum: nums.length ? Math.max(...nums) : 999
    }
  }

  if (allMatch(detectFrom, INT) || allMatch(pool, INT)) {
    const nums = pool.filter((s) => INT.test(s)).map(Number)
    // Mixed padded?
    if (pool.some((s) => /^-?0\d+/.test(s))) {
      const widths = pool.map((s) => s.replace(/^-/, '').length)
      return {
        kind: 'int-padded',
        samples: pool,
        minLen,
        maxLen,
        padWidth: Math.max(...widths),
        minNum: Math.min(...nums),
        maxNum: Math.max(...nums)
      }
    }
    return {
      kind: 'int',
      samples: pool,
      minLen,
      maxLen,
      minNum: Math.min(...nums),
      maxNum: Math.max(...nums)
    }
  }

  if (allMatch(detectFrom, FLOAT) || allMatch(pool, FLOAT)) {
    const nums = pool.map(Number)
    const decimals = Math.max(...pool.map(decimalPlaces), 1)
    return {
      kind: 'float',
      samples: pool,
      minLen,
      maxLen,
      minNum: Math.min(...nums),
      maxNum: Math.max(...nums),
      decimals
    }
  }

  // Phone-like (if mostly digits / separators)
  if (
    (sample && PHONE.test(sample) && /\d{7,}/.test(sample.replace(/\D/g, ''))) ||
    (allMatch(pool, PHONE) && pool.every((s) => s.replace(/\D/g, '').length >= 7))
  ) {
    return { kind: 'phone', samples: pool, minLen, maxLen }
  }

  if (allMatch(detectFrom, ALPHA) || allMatch(pool, ALPHA)) {
    return { kind: 'alpha', samples: pool, minLen, maxLen }
  }

  if (allMatch(detectFrom, ALNUM) || allMatch(pool, ALNUM)) {
    return { kind: 'alnum', samples: pool, minLen, maxLen }
  }

  return { kind: 'string', samples: pool, minLen, maxLen }
}

export function isDateKind(kind: InferredKind): boolean {
  return (
    kind === 'date-iso' ||
    kind === 'datetime-iso' ||
    kind === 'date-us' ||
    kind === 'date-eu' ||
    kind === 'date-slash-ymd'
  )
}

export function isNumericKind(kind: InferredKind): boolean {
  return (
    kind === 'int' ||
    kind === 'int-padded' ||
    kind === 'float' ||
    kind === 'currency' ||
    kind === 'percent'
  )
}

File: backend/app/services/patterns.py
"""Infer field formats from samples/history (port of Electron patterns.ts)."""
from __future__ import annotations

import re
from typing import Any

ISO_DATE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
ISO_DATETIME = re.compile(
    r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{1,3})?(Z|[+-]\d{2}:\d{2})?$"
)
DATE_US = re.compile(r"^(0?[1-9]|1[0-2])/(0?[1-9]|[12]\d|3[01])/\d{4}$")
DATE_EU = re.compile(r"^(0?[1-9]|[12]\d|3[01])/(0?[1-9]|1[0-2])/\d{4}$")
DATE_YMD_SLASH = re.compile(r"^\d{4}/(0?[1-9]|1[0-2])/(0?[1-9]|[12]\d|3[01])$")
EMAIL = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$")
UUID = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", re.I
)
PHONE = re.compile(r"^[+]?[\d\s().-]{7,20}$")
INT = re.compile(r"^-?\d+$")
FLOAT = re.compile(r"^-?\d+\.\d+$")
CURRENCY = re.compile(
    r"^(\$|€|£)?-?\d{1,3}(,\d{3})*(\.\d+)?$|^(\$|€|£)?-?\d+(\.\d+)?$"
)
PERCENT = re.compile(r"^-?\d+(\.\d+)?%$")
ALPHA = re.compile(r"^[A-Za-z]+$")
ALNUM = re.compile(r"^[A-Za-z0-9_-]+$")

DATE_KINDS = {
    "date-iso",
    "datetime-iso",
    "date-us",
    "date-eu",
    "date-slash-ymd",
}
NUMERIC_KINDS = {"int", "int-padded", "float"}


def is_date_kind(kind: str) -> bool:
    return kind in DATE_KINDS


def is_numeric_kind(kind: str) -> bool:
    return kind in NUMERIC_KINDS


def _unique(values: list[str]) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []
    for v in values:
        s = (v or "").strip()
        if not s or s in seen:
            continue
        seen.add(s)
        out.append(s)
    return out


def _all_match(samples: list[str], rx: re.Pattern) -> bool:
    return bool(samples) and all(rx.match(s) for s in samples)


def _decimals(s: str) -> int:
    m = re.search(r"\.(\d+)", s.replace("%", ""))
    return len(m.group(1)) if m else 0


def _parse_loose(s: str) -> float:
    return float(re.sub(r"[$€£,%]", "", s).replace(",", ""))


def detect_pattern(history: list[str], sample_value: str | None = None) -> dict[str, Any]:
    sample = (sample_value or "").strip() or None
    pool = _unique(([sample] if sample else []) + list(history))
    if not pool:
        return {"kind": "string", "samples": [], "minLen": 4, "maxLen": 12}

    detect_from = [sample] if sample else pool
    lens = [len(s) for s in pool]
    min_len, max_len = min(lens), max(lens)

    if _all_match(detect_from, re.compile(r"^(true|false)$", re.I)) or _all_match(
        pool, re.compile(r"^(true|false)$", re.I)
    ):
        return {"kind": "bool", "samples": pool, "minLen": 4, "maxLen": 5}

    if _all_match(detect_from, ISO_DATE) or (_all_match(pool, ISO_DATE) and not sample):
        return {"kind": "date-iso", "samples": pool, "minLen": 10, "maxLen": 10}
    if _all_match(detect_from, ISO_DATETIME) or _all_match(pool, ISO_DATETIME):
        return {"kind": "datetime-iso", "samples": pool, "minLen": 19, "maxLen": 30}

    if sample and DATE_US.match(sample) and not DATE_EU.match(sample):
        return {"kind": "date-us", "samples": pool, "minLen": 8, "maxLen": 10}
    if sample and DATE_EU.match(sample) and not sample.startswith(
        tuple("0123456789")[:0]
    ):
        day = int(sample.split("/")[0])
        if day > 12 or DATE_EU.match(sample):
            return {"kind": "date-eu", "samples": pool, "minLen": 8, "maxLen": 10}
    if _all_match(detect_from, DATE_YMD_SLASH) or _all_match(pool, DATE_YMD_SLASH):
        return {"kind": "date-slash-ymd", "samples": pool, "minLen": 8, "maxLen": 10}
    if _all_match(detect_from, DATE_US) or _all_match(pool, DATE_US):
        return {"kind": "date-us", "samples": pool, "minLen": 8, "maxLen": 10}
    if _all_match(detect_from, DATE_EU) or _all_match(pool, DATE_EU):
        return {"kind": "date-eu", "samples": pool, "minLen": 8, "maxLen": 10}

    if _all_match(detect_from, EMAIL) or _all_match(pool, EMAIL):
        return {"kind": "email", "samples": pool, "minLen": min_len, "maxLen": max_len}
    if _all_match(detect_from, UUID) or _all_match(pool, UUID):
        return {"kind": "uuid", "samples": pool, "minLen": 36, "maxLen": 36}

    def _safe_nums(values: list[str]) -> list[float]:
        out: list[float] = []
        for s in values:
            try:
                out.append(_parse_loose(s))
            except (ValueError, TypeError):
                continue
        return out or [0.0]

    if _all_match(detect_from, PERCENT) or _all_match(pool, PERCENT):
        src = [s for s in pool if PERCENT.match(s)] or detect_from
        nums = _safe_nums(src)
        return {
            "kind": "percent",
            "samples": pool,
            "minLen": min_len,
            "maxLen": max_len,
            "minNum": min(nums),
            "maxNum": max(nums),
            "decimals": max((_decimals(s) for s in src), default=0),
        }

    if _all_match(detect_from, CURRENCY) or _all_match(pool, CURRENCY):
        src = sample or pool[0]
        m = re.match(r"^(\$|€|£)", src)
        numeric_src = [s for s in pool if CURRENCY.match(s)] or detect_from
        nums = _safe_nums(numeric_src)
        return {
            "kind": "currency",
            "samples": pool,
            "minLen": min_len,
            "maxLen": max_len,
            "minNum": min(nums),
            "maxNum": max(nums),
            "decimals": max((_decimals(s) for s in numeric_src), default=2) or 2,
            "currencyPrefix": m.group(1) if m else "$",
            "useThousands": any("," in s for s in numeric_src),
        }

    if _all_match(detect_from, FLOAT) or _all_match(pool, FLOAT):
        src = [s for s in pool if FLOAT.match(s)] or detect_from
        nums = [float(s) for s in src]
        return {
            "kind": "float",
            "samples": pool,
            "minLen": min_len,
            "maxLen": max_len,
            "minNum": min(nums),
            "maxNum": max(nums),
            "decimals": max((_decimals(s) for s in src), default=2),
        }

    if _all_match(detect_from, INT) or _all_match(pool, INT):
        src = [s for s in pool if INT.match(s)] or detect_from
        nums = [int(s) for s in src]
        pad = all(s.lstrip("-").startswith("0") and len(s.lstrip("-")) > 1 for s in src)
        if pad or (sample and sample.lstrip("-").startswith("0") and len(sample) > 1):
            width = max(len(s.lstrip("-")) for s in src)
            return {
                "kind": "int-padded",
                "samples": pool,
                "minLen": min_len,
                "maxLen": max_len,
                "minNum": min(nums),
                "maxNum": max(nums),
                "padWidth": width,
            }
        return {
            "kind": "int",
            "samples": pool,
            "minLen": min_len,
            "maxLen": max_len,
            "minNum": min(nums),
            "maxNum": max(nums),
        }

    if _all_match(detect_from, PHONE) or _all_match(pool, PHONE):
        return {"kind": "phone", "samples": pool, "minLen": min_len, "maxLen": max_len}
    if _all_match(detect_from, ALPHA) or _all_match(pool, ALPHA):
        return {"kind": "alpha", "samples": pool, "minLen": min_len, "maxLen": max_len}
    if _all_match(detect_from, ALNUM) or _all_match(pool, ALNUM):
        return {"kind": "alnum", "samples": pool, "minLen": min_len, "maxLen": max_len}

    return {
        "kind": "string",
        "samples": pool,
        "minLen": min_len,
        "maxLen": max_len,
    }
